# 23

★  
引数にオブジェクト以外のものを渡した時にコンパイルエラーになって欲しい

```typescript
declare function foge<A = Record<string, unknown>>(
  a: A
): any

const a = foge(1)
```
↑ はコンパイルエラーになってくれない

```typescript
declare function foge<A extends Record<string, unknown> = Record<string, unknown>>(
  a: A
): any

const a = foge(1)
```
↑ こうすればコンパイルエラーになる。(型引数に `extends` を追加する。)

直感的には前者の方でもエラーになってくれそうなのに、そうでないのはなぜなんだろう。

★  
Promise<T>: Tはresolveの返却値を定義できる。rejectの値は常にany

```typescript
function iReturnPromiseAfter1Second(): Promise<string> {
    return new Promise((resolve) => {
        setTimeout(() => resolve("Hello world!!!!"), 1000);
    });
}

iReturnPromiseAfter1Second().then(r => console.log(r), e => console.log(e))
```
thenのあとの`r`はstringで推論されるが、`e`はanyになる



